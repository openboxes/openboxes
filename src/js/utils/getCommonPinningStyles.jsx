const getCommonPinningStyles = (
  column,
  flexWidth,
  isScreenWiderThanTable,
  isEmptyTableMessageOrLoading,
) => {
  const isPinned = column.getIsPinned();
  const isLastLeftPinnedColumn = isPinned === 'left' && column.getIsLastColumn('left');

  return {
    boxShadow: isLastLeftPinnedColumn ? '0 0 15px #00000040' : undefined,
    clipPath: isLastLeftPinnedColumn ? 'inset(0 -15px 0 0)' : undefined,
    marginRight: isLastLeftPinnedColumn ? '5px' : undefined,
    left: isPinned === 'left' ? `${column.getStart('left')}px` : undefined,
    position: isPinned && !isScreenWiderThanTable && !isEmptyTableMessageOrLoading && 'sticky',
    flex: flexWidth || column.getSize(),
    width: flexWidth || column.getSize(),
    zIndex: isPinned ? 1 : 0,
    background: isPinned && 'white',
  };
};

export default getCommonPinningStyles;
