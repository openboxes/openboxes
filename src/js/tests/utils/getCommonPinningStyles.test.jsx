import getCommonPinningStyles from 'utils/getCommonPinningStyles';

const mockColumn = {
  getIsPinned: jest.fn(),
  getIsLastColumn: jest.fn(),
  getStart: jest.fn(),
  getSize: jest.fn(),
};

describe('getCommonPinningStyles', () => {
  const mockParams = {
    column: mockColumn,
    flexWidth: undefined,
    isScreenWiderThanTable: false,
    dataLength: 10,
    loading: false,
  };

  beforeEach(() => {
    jest.clearAllMocks();
    mockColumn.getIsPinned.mockReturnValue(false);
    mockColumn.getIsLastColumn.mockReturnValue(false);
    mockColumn.getStart.mockReturnValue(0);
    mockColumn.getSize.mockReturnValue(100);
  });

  it('should match snapshot for sample params', () => {
    const result = getCommonPinningStyles(
      mockParams.column,
      mockParams.flexWidth,
      mockParams.isScreenWiderThanTable,
      mockParams.dataLength,
      mockParams.loading,
    );
    expect(result)
      .toMatchSnapshot();
  });

  it('returns styles for non-pinned column', () => {
    const result = getCommonPinningStyles(
      mockParams.column,
      mockParams.flexWidth,
      mockParams.isScreenWiderThanTable,
      mockParams.dataLength,
      mockParams.loading,
    );
    expect(result)
      .toEqual({
        boxShadow: undefined,
        clipPath: undefined,
        marginRight: undefined,
        left: undefined,
        position: false,
        flex: 100,
        width: 100,
        zIndex: 0,
        background: false,
      });
  });

  it('returns styles for last left-pinned column', () => {
    mockColumn.getIsPinned.mockReturnValue('left');
    mockColumn.getIsLastColumn.mockReturnValue(true);
    mockColumn.getStart.mockReturnValue(50);
    const result = getCommonPinningStyles(
      mockParams.column,
      mockParams.flexWidth,
      mockParams.isScreenWiderThanTable,
      mockParams.dataLength,
      mockParams.loading,
    );
    expect(result)
      .toEqual({
        boxShadow: '0 0 15px #00000040',
        clipPath: 'inset(0 -15px 0 0)',
        marginRight: '5px',
        left: '50px',
        position: 'sticky',
        flex: 100,
        width: 100,
        zIndex: 1,
        background: 'white',
      });
  });

  it('disables sticky position when conditions are not met', () => {
    mockColumn.getIsPinned.mockReturnValue('left');
    const result1 = getCommonPinningStyles(
      mockParams.column,
      mockParams.flexWidth,
      true,
      mockParams.dataLength,
      mockParams.loading,
    );
    const result2 = getCommonPinningStyles(
      mockParams.column,
      mockParams.flexWidth,
      mockParams.isScreenWiderThanTable,
      0,
      mockParams.loading,
    );
    const result3 = getCommonPinningStyles(
      mockParams.column,
      mockParams.flexWidth,
      mockParams.isScreenWiderThanTable,
      mockParams.dataLength,
      true,
    );
    expect(result1.position)
      .toBeFalsy();
    expect(result2.position)
      .toBeFalsy();
    expect(result3.position)
      .toBeFalsy();
  });

  it('handles undefined column methods safely', () => {
    const undefinedColumn = {
      getIsPinned: jest.fn()
        .mockReturnValue(undefined),
      getIsLastColumn: jest.fn()
        .mockReturnValue(undefined),
      getStart: jest.fn()
        .mockReturnValue(undefined),
      getSize: jest.fn()
        .mockReturnValue(undefined),
    };
    const result = getCommonPinningStyles(
      undefinedColumn,
      mockParams.isScreenWiderThanTable,
      mockParams.dataLength,
      mockParams.loading,
    );
    expect(result)
      .toEqual({
        boxShadow: undefined,
        clipPath: undefined,
        marginRight: undefined,
        left: undefined,
        position: undefined,
        flex: undefined,
        width: undefined,
        zIndex: 0,
        background: undefined,
      });
  });
});
