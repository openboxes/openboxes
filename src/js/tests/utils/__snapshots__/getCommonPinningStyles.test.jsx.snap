// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCommonPinningStyles should match snapshot for sample params 1`] = `
{
  "background": false,
  "boxShadow": undefined,
  "clipPath": undefined,
  "flex": 100,
  "left": undefined,
  "marginRight": undefined,
  "position": false,
  "width": 100,
  "zIndex": 0,
}
`;
